src/rbtree.rs

```diff
@@ -95,422 +154,929 @@ impl fmt::Debug for Link {
+    if let Some(parent) = link_ops.parent(ptr) {
+        if is_left_child(link_ops, ptr, parent) {
+            link_ops.set_left(parent, Some(y));
         } else {
-            self.parent().set_right(y);
+            link_ops.set_right(parent, Some(y));
         }
-        y.set_left(self);
-        self.set_parent(y);
+    } else {
+        *root = Some(y);
     }
+    link_ops.set_left(y, Some(ptr));
+    link_ops.set_parent(ptr, Some(y));
+}
 
-    unsafe fn rotate_right(self, root: &mut NodePtr) {
-        let y = self.left();
-        self.set_left(y.right());
-        if !self.left().is_null() {
-            self.left().set_parent(self);
-        }
-        y.set_parent(self.parent());
-        if self.parent().is_null() {
-            *root = y;
-        } else if self.is_left_child() {
-            self.parent().set_left(y);
+unsafe fn rotate_right<T: RBTreeOps>(
+    link_ops: &mut T,
+    ptr: T::LinkPtr,
+    root: &mut Option<T::LinkPtr>,
+) {
+    let y = link_ops
+        .left(ptr)
+        .unwrap_or_else(|| hint::unreachable_unchecked());
+    link_ops.set_left(ptr, link_ops.right(y));
+    if let Some(left) = link_ops.left(ptr) {
+        link_ops.set_parent(left, Some(ptr));
+    }
+    link_ops.set_parent(y, link_ops.parent(ptr));
+    if let Some(parent) = link_ops.parent(ptr) {
+        if is_left_child(link_ops, ptr, parent) {
+            link_ops.set_left(parent, Some(y));
         } else {
-            self.parent().set_right(y);
-        }
-        y.set_right(self);
-        self.set_parent(y);
-    }
-
-    // This code is based on the red-black tree implementation in libc++
-    unsafe fn post_insert(self, root: &mut NodePtr) {
-        let mut x = self;
-        while !x.parent().is_null() && x.parent().color() == Color::Red {
-            if x.parent().is_left_child() {
-                let y = x.parent().parent().right();
-                if !y.is_null() && y.color() == Color::Red {
-                    x = x.parent();
-                    x.set_color(Color::Black);
-                    x = x.parent();
-                    if x.parent().is_null() {
-                        x.set_color(Color::Black);
-                    } else {
-                        x.set_color(Color::Red);
-                    }
-                    y.set_color(Color::Black);
+            link_ops.set_right(parent, Some(y));
+        }
+    } else {
+        *root = Some(y);
+    }
+    link_ops.set_right(y, Some(ptr));
+    link_ops.set_parent(ptr, Some(y));
+}
+
+// This code is based on the red-black tree implementation in libc++
+unsafe fn post_insert<T: RBTreeOps>(
+    link_ops: &mut T,
+    ptr: T::LinkPtr,
+    root: &mut Option<T::LinkPtr>,
+) {
+    let mut x = ptr;
+    while !link_ops.parent(x).is_none()
+        && link_ops.color(
+            link_ops
+                .parent(x)
+                .unwrap_or_else(|| hint::unreachable_unchecked()),
+        ) == Color::Red
+    {
```

Comment:
`let grandparent = link_ops.parent(parent).unwrap_unchecked()`

This is safe since the root of the tree must be black, and `parent` can't be the the root due to the check above.